server/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java

```diff
@@ -223,4 +303,21 @@ public String toString() {
     public boolean isEmpty() {
         return shards.isEmpty();
     }
+
+    private boolean invariant() {
```

Comment:
Please would you also assert this on entry to the methods that change the state (`add()`, `remove()` and `update()`)